File: 1d_dynamic_programming/PartitionEqualSubsetSum.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/*
Run in the neetcode text editor.
*/
public class PartitionEqualSubsetSum {
 
  // Dynamic Programming solution (best) | Time: O(n * target), Space: O(target),
  // where n is the length of the array nums and target is the sum of array
  // elements divided by 2.
  public boolean canPartition(int[] nums) {
    if (Arrays.stream(nums).sum() % 2 != 0){
      return false;
    }

    Set<Integer> dp = new HashSet<>();
    dp.add(0);
    int target = Arrays.stream(nums).sum() / 2;

    for (int i = nums.length - 1; i >= 0; i--){
    Set<Integer> nextDP = new HashSet<>();
      for (int t : dp){
        if (t + nums[i] == target){
          return true;
        }
        nextDP.add(t + nums[i]);
        nextDP.add(t);
      }
      dp = nextDP;
    }
    return false;
  }

}

File: 1d_dynamic_programming/partition_equal_subset_sum.py
from typing import List

"""
Run in the neetcode text editor.
"""
class Solution:

  # Dynamic Programming solution (best) | Time: O(n * target), Space: O(target),
  # where n is the length of the array nums and target is the sum of array
  # elements divided by 2.
  def canPartition(self, nums: List[int]) -> bool:
    if sum(nums) % 2:
      return False

    dp = set()
    dp.add(0)
    target = sum(nums) // 2

    for i in range(len(nums) - 1, -1, -1):
      nextDP = set()
      for t in dp:
        if (t + nums[i] == target):
          return True
        nextDP.add(t + nums[i])
        nextDP.add(t)
      dp = nextDP
    return False